### ISSUE TITLE
Parser error for array value

### ISSUE THREAD

**Issue by kwin** [boothen/Json-Eclipse-Plugin#9]
The following (valid) JSON leads to errors
`{
    "validatorArguments" : ["regex=^\\\p{Upper}+$"]
}`

**Comment by msteiger** [comment]
It looks correct to me, too. However, jsonlint says no:
http://jsonlint.com/

Using two backslashes instead of three makes it work again. Can you confirm this? I would think that escaping in JSON values is applied even if you don't want to.

**Comment by kwin** [comment]
Yes, jsonlint says it is invalid, but http://www.freeformatter.com/json-validator.html says it is valid. Using two backslashes would modify the meaning here. I don't see why three backslashes should not be allowed here, because you are allowed to escape everything with a backslash in JSON.

At least it says so in https://tools.ietf.org/html/rfc4627, section 2.5
> Any character may be escaped.

**Comment by msteiger** [comment]
I agree - it should not fail. The error seems to be unrelated to the array parsing, but rather be caused by the triple escaping. I created the parser using the following grammer: https://github.com/boothen/Json-Eclipse-Plugin/blob/master/jsonedit-model/grammar/JSON.g4

It describes how the escaping is done in the `ESC` fragment.

**Comment by kwin** [comment]
Actually even though the escape sequences in that grammar are the mandatory ones, everything else may be escaped as well.

**Comment by msteiger** [comment]
> .. everything else may be escaped as well.

Well, the answer is complex, it seems. The standard ECMA 404 says **no** (see https://www.json.org), but the RFC family says **yes**: https://tools.ietf.org/html/rfc8259#page-8

The ANTLR grammar follows ECMA, that's why you get the error. Maybe we can be a little lenient here and flag it as a warning ... would that work for you?

**Comment by diggernet** [comment]
Hi Martin,

I think the two standards actually agree with each other.  The RFC text:

>   Any character may be escaped.  If the character is in the Basic
>   Multilingual Plane (U+0000 through U+FFFF), then it may be
>   represented as a six-character sequence: a reverse solidus, followed
>   by the lowercase letter u, followed by four hexadecimal digits that
>   encode the character's code point.  
>
>   Alternatively, there are two-character sequence escape
>   representations of some popular characters.

It follows that with a BNF spec showing 8 specific two-character sequences.
So I think what it's saying is that any character may be escaped *with the \u#### sequence*, and just those 8 can be escaped with the two-char sequence.  By that reading, \p is not allowed.

Whether that is considered an error or a warning, I don't have a strong opinion on.  I suppose it depends on how the JSON library someone is using reacts to those sequences.  So perhaps it could be a configurable error/warning.

But what brought me here in the first place is \r and \t.  Those two escapes are clearly defined in both specs as legal, but are getting flagged as invalid.  It'd be nice to have them added to the Pattern in StringProblemFinder.checkEscaping().

Thanks,
Dave

**Comment by msteiger** [comment]
I think that it's the first line of your quote that describes the difference. In RFC 8259 *any* char. may be escaped while ECMA allows only the 8+1 you describe.

So I implemented this as a warning for now, but I guess those 8+1 should not be flagged in any case. Thanks for pointing this out. Will have a look after Easter.

**Comment by diggernet** [comment]
Yup, 1.1.0.

**Comment by diggernet** [comment]
Looks good, thanks!
